### PR TITLE
feat: Add agent API's proto

### PR DESCRIPTION
**Pull request by EstebanBAR0N**
## Overall explanation of your work :
Add a proto file in order to specify the agent API routes and types

## Tasks lists :

- [x] Define .proto file
    - [x] Define structures
    - [x] Define enumerations
    - [x] Define services

### CODE REVIEW

File: docs/proposals/agent.proto [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// Represents the different Type of a workflow
+enum Type {
+    CONTAINER = 0;
+}
+
+service SchedulerService {
+    rpc InstanceCreate (Instance) returns (google.protobuf.Empty) {}
+    rpc InstanceUpdate (InstanceAction) returns (google.protobuf.Empty) {}
```
Comment (sameo) [inline docs/proposals/agent.proto:52]: I'd prefer for it to be called `InstanceManagement`. Even better, I'd prefer to have one endpoint for each action, which are stop, destroy and kill. You don't need start, the instance starts when it's being created.

File: docs/proposals/agent.proto [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	string id = 1;
+	Action action = 2;
+}
+
+// Represents the different Instance actions possible
+enum Action {
+	START = 0;
+	STOP = 1;
+	DESTROY = 2;
+	KILL = 3;
```
[the remark this answers is not in the document]
Comment (EstebanBAR0N) [reply docs/proposals/agent.proto:42]: Kill is used to force stop a container while destroy removes a container (it's the equivalent of a docker container rm). 
Isn't signal a little bit too generic in this case ?

File: docs/proposals/agent.proto [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// Represents a summary of all necessary resources
+message ResourceSummary {
+	int cpu = 1;
+	int memory = 2;
+	int disk = 3;
+}
+
+// Represent the maximum/usage of a Instance or a Node
```
[the remark this answers is not in the document]
Comment (EstebanBAR0N) [reply docs/proposals/agent.proto:24]: Usage specify the resource (ram, cpu and disk) currently used by the instance. 
We can imagine creating dashboards with those metrics.

[the remark this answers is not in the document]
Comment (EstebanBAR0N) [reply docs/proposals/agent.proto:24]: sorry, we just understood, we will correct it

File: docs/proposals/agent.proto [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+message InstanceStatus {
+  string id = 1;
+  Status status = 2;
+  string description = 3;
+}
+
+
+// Represent the maximum resource usage of an Instance
+message MaxResource {
+	int32 cpu = 1;
```
Comment (sameo) [inline docs/proposals/agent.proto:48]: nit: Please fix the indentation.

File: docs/proposals/agent.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (sameo) [reply docs/proposals/agent.proto]: `signal`will give you flexibility and you may want to actually send specific signals to your containers. From a UX perspective, you can implement your kill semantic by sending `Signal(KILL)`. So I'd advice to replace kill with signal.

Also I would actually get rid of DESTROY. Having that action means you want to support instance restart, which means potentially having to keep and maintain a more complex state machine. Also, I don't expect a cluster user to be used to stop and then destroy an instance. Implicitly you expect instances to be removed when being stopped.

**Review by sameo — COMMENTED** [review]
Almost there, a couple of comments to address and then we can merge that PR.

File: docs/proposals/agent.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexis-langlet) [reply docs/proposals/agent.proto]: Ok, thank you. 
We choose to use signal as suggested. 
We will only use stop and kill (that have the same meaning that SIGSTOP and SIGKILL), both will remove containers after finishing. 
In the same way, we decided to remove start service, a newly created container will automatically be started.

**Pull request merged**